### ISSUE TITLE
KeyboardEvent.key may be undefined

### ISSUE THREAD

**Issue by PinkChampagne17** [JohannesKlauss/react-hotkeys-hook#834]
First, I'm sorry for not being able to reproduce this.

I got some feedback that react-hotkeys-hook may throw Type Error: read properties of undefined (reading 'toLowerCase' or 'toLocaleLowerCase').

This situation mainly occurs in Chrome 107 on Mac OS / Windows.

Stack trace 1:
https://github.com/JohannesKlauss/react-hotkeys-hook/blob/cc8aeed82c4bf1963afacc47efb59b2a5a8e077f/src/useHotkeys.ts#L81

Stack trace 2:
https://github.com/JohannesKlauss/react-hotkeys-hook/blob/cc8aeed82c4bf1963afacc47efb59b2a5a8e077f/src/parseHotkeys.ts#L24

https://github.com/JohannesKlauss/react-hotkeys-hook/blob/9316dcc62e3432f9fb0999170fcf72c1ac66363c/src/isHotkeyPressed.ts#L24

**Comment by JohannesKlauss** [comment]
thanks for raising that issue. I'll try to come up with a fix.

**Comment by RickLucassen** [comment]
Hello,

We are currently running into the same issue. We notice this issue popping up in combination with the chrome auto-fill mechanism. At that point the key seems to be undefined.

![image](https://user-images.githubusercontent.com/44491854/201369266-cf8e2420-e125-4c86-8234-48736a4716f4.png)

**Comment by JohannesKlauss** [comment]
Hey @RickLucassen thanks for the additional info. Do you happen to have a reproduction sandbox at your hand?

**Comment by RickLucassen** [comment]
Sure, here you go: https://codesandbox.io/s/affectionate-wescoff-3ux0g8
If you open this in chrome and use the autofill mechanism to autofill something in one of the inputs it generated the error mentioned above.

**Comment by linonetwo** [comment]
I wonder why we need a listener on document, why not only listen on \<HotkeysProvider ?

```
(function() {
  if (typeof window !== "undefined") {
    window.addEventListener("DOMContentLoaded", function() {
      document.addEventListener("keydown", function(e5) {
        pushToCurrentlyPressedKeys(e5.key);
      });
      document.addEventListener("keyup", function(e5) {
        removeFromCurrentlyPressedKeys(e5.key);
      });
    });
  }
})();
```

**Comment by JohannesKlauss** [comment]
Well, you cannot "listen" to the HotkeysProvider, since it's a React Context Provider Component. The functionality of `isHotkeyPressed` should work without the need of having to use the HotkeysProvider component.

Also I am not sure what this has to do with the issue?

**Comment by linonetwo** [comment]
Oh, just tiddlywiki clone the event, and strip out the "key" property, and propergate it to the document, and catch by this snippet of code, causing the error if this issue's title.

I have fixed this in the tiddlywiki side for my project. But I still hope in react-hotkeys-hook v4 we can listen on specific dom node instead of the document.

**Comment by JohannesKlauss** [comment]
I don't quite understand. You can listen to a specific dom node with the returned ref of the `useHotkeys` hook, like this:

```tsx
function ScopedHotkey() {
  const [count, setCount] = useState(0)
  const ref = useHotkeys('shift+a', () => setCount(prevCount => prevCount + 1))

  return (
    <>
      <p>The count is {count}. Click anywhere expect for the button to disable the hotkey.</p>
      <button ref={ref}>
        Click me to enable the hotkey
      </button>
    </>
  )
}
```

The source code you pointed to is only related to `isHotkeyPressed` which is an imperative check of all the pressed down keys in a specific moment.

**Comment by linonetwo** [comment]
Sorry, I tried this again and found if I use ref, the error is from here:

```typescript
var useSafeLayoutEffect = typeof window !== "undefined" ? import_react4.useLayoutEffect : import_react4.useEffect;
function useHotkeys(keys, callback, options, dependencies) {
  var ref = (0, import_react4.useRef)(null);
  var _useRef = (0, import_react4.useRef)(/* @__PURE__ */ new Set()), pressedDownKeys = _useRef.current;
  var _options = !(options instanceof Array) ? options : !(dependencies instanceof Array) ? dependencies : void 0;
  var _deps = options instanceof Array ? options : dependencies instanceof Array ? dependencies : [];
  var cb = (0, import_react4.useCallback)(callback, [].concat(_deps));
  var memoisedOptions = useDeepEqualMemo(_options);
  var _useHotkeysContext = useHotkeysContext(), enabledScopes = _useHotkeysContext.enabledScopes;
  var proxy = useBoundHotkeysProxy();
  useSafeLayoutEffect(function() {
    if ((memoisedOptions == null ? void 0 : memoisedOptions.enabled) === false || !isScopeActive(enabledScopes, memoisedOptions == null ? void 0 : memoisedOptions.scopes)) {
      return;
    }
    var listener = function listener2(e5) {
      var _e$target;
      if (isKeyboardEventTriggeredByInput(e5) && !isHotkeyEnabledOnTag(e5, memoisedOptions == null ? void 0 : memoisedOptions.enableOnFormTags)) {
        return;
      }
      if (ref.current !== null && document.activeElement !== ref.current && !ref.current.contains(document.activeElement)) {
        stopPropagation(e5);
        return;
      }
      if ((_e$target = e5.target) != null && _e$target.isContentEditable && !(memoisedOptions != null && memoisedOptions.enableOnContentEditable)) {
        return;
      }
      parseKeysHookInput(keys, memoisedOptions == null ? void 0 : memoisedOptions.splitKey).forEach(function(key) {
        var _hotkey$keys;
        var hotkey = parseHotkey(key, memoisedOptions == null ? void 0 : memoisedOptions.combinationKey);
        if (isHotkeyMatchingKeyboardEvent(e5, hotkey, pressedDownKeys) || (_hotkey$keys = hotkey.keys) != null && _hotkey$keys.includes("*")) {
          maybePreventDefault(e5, hotkey, memoisedOptions == null ? void 0 : memoisedOptions.preventDefault);
          if (!isHotkeyEnabled(e5, hotkey, memoisedOptions == null ? void 0 : memoisedOptions.enabled)) {
            stopPropagation(e5);
            return;
          }
          cb(e5, hotkey);
        }
      });
    };
    var handleKeyDown = function handleKeyDown2(event) {
err-> pressedDownKeys.add(event.key.toLowerCase());
      if ((memoisedOptions == null ? void 0 : memoisedOptions.keydown) === void 0 && (memoisedOptions == null ? void 0 : memoisedOptions.keyup) !== true || memoisedOptions != null && memoisedOptions.keydown) {
        listener(event);
      }
    };
```

**Comment by linonetwo** [comment]
So the ref is working, but acturally I was typing in a textarea outside of this element, so I was confused.

**Comment by JohannesKlauss** [comment]
Should be fixed in 4.0.5
